### PR TITLE
[AARCH64 Automation] [WIP] More reliable way to activate sol session

### PR DESCRIPTION
**Pull request by waynechen55**
1.Before activating sol session, bmc needs to be reset and then deactivated. This way can help guarantee active sol session is always available anytime during test run.
2.The while loop here is to make sure the least amount of time elapsed before re-activate sol session following bmc comes back online.
3.New global variable BMC_IP has been defined. If to be used, it should be added into MACHINE.
4.No ARCH dependent code has been added into this change. If needed, it should only be applied to aarch64.

- Related ticket: aarch64 ipmi backend
- Needles: n/a
- Verification run: http://10.162.187.154/tests/349

### CODE REVIEW

**Comment by Martchus** [comment]
You need to run `tools/tidy`.

**Comment by waynechen55** [comment]
> You need to run `tools/tidy`.

Done

File: consoles/sshXtermIPMI.pm [hunk 1]
```diff
@@ -22,6 +22,7 @@ use testapi 'get_required_var';
 require IPC::System::Simple;
 use autodie ':all';
 use File::Which;
+use testapi 'get_var';
```
Comment (okurz) [inline consoles/sshXtermIPMI.pm]: Please add that in line 21. However IMHO using the test API from within the consoles is bad design

File: consoles/sshXtermIPMI.pm [hunk 2]
```diff
@@ -31,11 +32,40 @@ sub activate {
...
-    my @command = $self->backend->ipmi_cmdline;
-    push(@command, qw(sol activate));
-    my $serial = $self->{args}->{serial};
+    my $cstr              = "";
+    my @command_reset_tmp = $self->backend->ipmi_cmdline;
+    push(@command_reset_tmp, qw(mc reset cold));
+    my $command_reset = join(' ', @command_reset_tmp);
+
+    my @command_sleep_tmp = "sleep";
+    push(@command_sleep_tmp, qw(15));
```
Comment (okurz) [inline consoles/sshXtermIPMI.pm]: Why do we need a sleep of 15? I assume it is seconds, 15 seconds would be a lot of time wasted

File: consoles/sshXtermIPMI.pm [hunk 3]
```diff
@@ -31,11 +32,40 @@ sub activate {
...
+    my $command_sleep = join(' ', @command_sleep_tmp);
+
+    my $bmcipaddr = get_var('BMC_IP');
```
Comment (okurz) [inline consoles/sshXtermIPMI.pm]: Can we not work with hostnames? Why need to work with IPs?

File: consoles/sshXtermIPMI.pm [hunk 4]
```diff
@@ -31,11 +32,40 @@ sub activate {
...
+    if ($bmcipaddr eq '') {
+        die "BMC IP address is not defined.";
+    }
+    my @command_ping_tmp = "(while true; do /usr/bin/ping -c1";
```
Comment (okurz) [inline consoles/sshXtermIPMI.pm]: What if the ping never succeeds? How about using a for loop to abort after a certain amount of retries

File: consoles/sshXtermIPMI.pm (hunk 1 above)
Comment (waynechen55) [reply consoles/sshXtermIPMI.pm]: I will think whether this can be avoided.

File: consoles/sshXtermIPMI.pm (hunk 2 above)
Comment (waynechen55) [reply consoles/sshXtermIPMI.pm]: To make sure bmc already offline before starting ping.

File: consoles/sshXtermIPMI.pm (hunk 3 above)
Comment (waynechen55) [reply consoles/sshXtermIPMI.pm]: Yes. There is already IPMI_HOSTNAME. Will change this.

File: consoles/sshXtermIPMI.pm (hunk 4 above)
Comment (waynechen55) [reply consoles/sshXtermIPMI.pm]: Will improve.

File: consoles/sshXtermIPMI.pm (hunk 2 above)
Comment (okurz) [reply consoles/sshXtermIPMI.pm]: Yes, clearly. My question is: Can you ensure that with a proper check, not sleeping?

**Comment by waynechen55** [comment]
Need more time to figure best solution out.

**Comment by waynechen55** [comment]
I will simplify implementation by moving the large patch of code out of activate function step by step.

**Comment by okurz** [comment]
@alice-suse this looks like going in the direction of what we would could use to reset the MC from tests properly

**Comment by waynechen55** [comment]
> @alice-suse this looks like going in the direction of what we would could use to reset the MC from tests properly

https://progress.opensuse.org/issues/36027
I read it through. It seems that we both share something in common although having been tried to solve different problems if say accurately. Virtualization PRJ2 test suites require multiple times of host power reset, during which sol session can be interrupted or lost somehow. The reasons behind this could be dozens, which I think are already covered in your discussion. So I tried to do "mc reset" before each time of host reboot to put "mc" into its original state, which I think can help increase the success rate when test suites switch back to sol console to capture sol session output. We converged in this point literally. As far as the specific implementation is concerned:
1) If you finally do it  jenkin way or openqa-not-related way, we do not have too much in common anymore.
2) If you finally implement the "mc reset" or similar function abstractly in a separate module or class or subroutine, which then is called upon or propagated into other test modules, I can take advantage of that as well. The only difference is I may need to insert it right just before every times of host reboot.

**Comment by waynechen55** [comment]
> @okurz @waynechen55 , Yes, I agree with you both mostly. The purpose is to do 'mc reset', the solutions can be slightly different according to where(backend or test?) and when(at the beginning of every test or every build or specific points like before reboot or when unstability detected) to do it.
> But the, core code will always be "do mc reset and wait it finish successfully or retry within given times".
> 
> Let's try the agreed on solution in poo#36027 first, if it can not achieve our target to be on stable ipmi through the tests, or it hurts stability with so repeated reset, then let's decide the next step.

I think flexibility is necessary, which then can be leveraged by other test suites or extended into other scenarios, whatever tests, builds or specific points. Just like the openQA itself.

**Comment by alice-suse** [comment]
@waynechen55 , the proposal to do mc reset from backend is merged, please refer to https://github.com/os-autoinst/os-autoinst/pull/1078. I guess this can meet your requirement for aarch64 :-)

**Comment by okurz** [comment]
With https://github.com/os-autoinst/os-autoinst/pull/1232 we should have a better approach to handle consoles that should be persistent. The idea mentioned in this PR could be still valid in case of recovering from instabilities but as there has not been any update by you within the last 10 months I assume you do not plan to continue. Please reopen if I am wrong.

**Comment by waynechen55** [comment]
> With #1232 we should have a better approach to handle consoles that should be persistent. The idea mentioned in this PR could be still valid in case of recovering from instabilities but as there has not been any update by you within the last 10 months I assume you do not plan to continue. Please reopen if I am wrong.

You can close. I will re-implement it better if necessary.

**Pull request closed without merging**